File: workers/email-router/src/protocol.ts
export const PROTOCOL_VERSION = "v1";
export const KEY_BYTES = 32;
export const TAG_LENGTH = 8;
export const MAX_LABEL_BYTES = 48;

const CONTEXT = "mailias/v1";
const TAG_ALPHABET = "023456789abcdefghjkmnpqrstuvwxyz";
const ASCII_ENCODER = new TextEncoder();

export interface VerifiedAlias {
  label: string;
  address: string;
}

export function normalizeLabel(input: string): string {
  if (input.length === 0) {
    throw new Error("the label must not be empty");
  }
  if (!isAscii(input) || ASCII_ENCODER.encode(input).length > MAX_LABEL_BYTES) {
    throw new Error("the label is not valid ASCII or is too long");
  }

  const label = input.toLowerCase();
  if (!/^[a-z0-9]+(?:-[a-z0-9]+)*$/.test(label)) {
    throw new Error("the label has invalid syntax");
  }
  return label;
}

export function normalizeDomain(input: string): string {
  if (
    input.length === 0 ||
    input.length > 253 ||
    !isAscii(input) ||
    input.endsWith(".")
  ) {
    throw new Error("the domain is invalid");
  }

  const domain = input.toLowerCase();
  const valid = domain
    .split(".")
    .every(
      (part) =>
        part.length > 0 &&
        part.length <= 63 &&
        !part.startsWith("-") &&
        !part.endsWith("-") &&
        /^[a-z0-9-]+$/.test(part),
    );
  if (!valid) {
    throw new Error("the domain is invalid");
  }
  return domain;
}

export function decodeKey(encoded: string): Uint8Array {
  if (!/^[A-Za-z0-9_-]{43}$/.test(encoded)) {
    throw new Error("the key is not canonical unpadded Base64URL");
  }
  const base64 = encoded.replaceAll("-", "+").replaceAll("_", "/") + "=";
  let binary: string;
  try {
    binary = atob(base64);
  } catch {
    throw new Error("the key is not valid Base64URL");
  }
  const key = Uint8Array.from(binary, (character) => character.charCodeAt(0));
  if (key.length !== KEY_BYTES) {
    throw new Error("the key must decode to exactly 32 bytes");
  }
  return key;
}

export async function generateTag(
  encodedKey: string,
  domainInput: string,
  labelInput: string,
): Promise<string> {
  const domain = normalizeDomain(domainInput);
  const label = normalizeLabel(labelInput);
  const keyBytes = decodeKey(encodedKey);
  const keyData = new Uint8Array(keyBytes.byteLength);
  keyData.set(keyBytes);
  const message = ASCII_ENCODER.encode(`${CONTEXT}\0${domain}\0${label}`);
  const key = await crypto.subtle.importKey(
    "raw",
    keyData,
    { name: "HMAC", hash: "SHA-256" },
    false,
    ["sign"],
  );
  const digest = new Uint8Array(await crypto.subtle.sign("HMAC", key, message));
  return encodeTag(digest.subarray(0, 5));
}

export async function generateAddress(
  encodedKey: string,
  domainInput: string,
  labelInput: string,
): Promise<string> {
  const domain = normalizeDomain(domainInput);
  const label = normalizeLabel(labelInput);
  const tag = await generateTag(encodedKey, domain, label);
  return `${label}-${PROTOCOL_VERSION}-${tag}@${domain}`;
}

export async function verifyAddress(
  encodedKey: string,
  expectedDomainInput: string,
  addressInput: string,
): Promise<VerifiedAlias | null> {
  const expectedDomain = normalizeDomain(expectedDomainInput);
  if (!isAscii(addressInput)) {
    return null;
  }

  const address = addressInput.toLowerCase();
  const separator = address.lastIndexOf("@");
  if (separator <= 0 || separator !== address.indexOf("@")) {
    return null;
  }
  const localPart = address.slice(0, separator);
  const domainInput = address.slice(separator + 1);

  let domain: string;
  try {
    domain = normalizeDomain(domainInput);
  } catch {
    return null;
  }
  if (domain !== expectedDomain) {
    return null;
  }

  const versionSeparator = localPart.lastIndexOf("-v1-");
  if (versionSeparator <= 0) {
    return null;
  }
  const labelInput = localPart.slice(0, versionSeparator);
  const tag = localPart.slice(versionSeparator + "-v1-".length);

  let label: string;
  try {
    label = normalizeLabel(labelInput);
  } catch {
    return null;
  }
  if (
    tag.length !== TAG_LENGTH ||
    ![...tag].every((character) => TAG_ALPHABET.includes(character))
  ) {
    return null;
  }

  const expectedTag = await generateTag(encodedKey, expectedDomain, label);
  if (!constantTimeEqual(expectedTag, tag)) {
    return null;
  }

  return {
    label,
    address: `${label}-${PROTOCOL_VERSION}-${expectedTag}@${expectedDomain}`,
  };
}

function encodeTag(bytes: Uint8Array): string {
  if (bytes.length !== 5) {
    throw new Error("a mailias tag requires exactly five bytes");
  }

  let value = 0;
  for (const byte of bytes) {
    value = value * 256 + byte;
  }

  const output = new Array<string>(TAG_LENGTH);
  for (let index = TAG_LENGTH - 1; index >= 0; index -= 1) {
    output[index] = TAG_ALPHABET[value % 32];
    value = Math.floor(value / 32);
  }
  return output.join("");
}

function constantTimeEqual(left: string, right: string): boolean {
  if (left.length !== right.length) {
    return false;
  }
  let difference = 0;
  for (let index = 0; index < left.length; index += 1) {
    difference |= left.charCodeAt(index) ^ right.charCodeAt(index);
  }
  return difference === 0;
}

function isAscii(input: string): boolean {
  return /^[\x00-\x7f]*$/.test(input);
}

File: crates/mailias-core/src/protocol.rs
use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine as _};
use hmac::{Hmac, Mac};
use sha2::Sha256;
use subtle::ConstantTimeEq;

use crate::ProtocolError;

pub const PROTOCOL_VERSION: &str = "v1";
pub const KEY_BYTES: usize = 32;
pub const TAG_BYTES: usize = 8;
pub const MAX_LABEL_BYTES: usize = 48;

const CONTEXT: &[u8] = b"mailias/v1";
const TAG_ALPHABET: &[u8; 32] = b"023456789abcdefghjkmnpqrstuvwxyz";

type HmacSha256 = Hmac<Sha256>;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct VerifiedAlias {
    pub label: String,
    pub address: String,
}

pub fn encode_key(key: &[u8; KEY_BYTES]) -> String {
    URL_SAFE_NO_PAD.encode(key)
}

pub fn decode_key(encoded: &str) -> Result<[u8; KEY_BYTES], ProtocolError> {
    let decoded = URL_SAFE_NO_PAD
        .decode(encoded)
        .map_err(|_| ProtocolError::InvalidKeyEncoding)?;
    decoded
        .try_into()
        .map_err(|_| ProtocolError::InvalidKeyLength)
}

pub fn normalize_label(input: &str) -> Result<String, ProtocolError> {
    if input.is_empty() {
        return Err(ProtocolError::EmptyLabel);
    }
    if input.len() > MAX_LABEL_BYTES {
        return Err(ProtocolError::LabelTooLong);
    }
    if !input.is_ascii() {
        return Err(ProtocolError::InvalidLabel);
    }

    let label = input.to_ascii_lowercase();
    if label.split('-').any(|part| {
        part.is_empty()
            || !part
                .bytes()
                .all(|byte| byte.is_ascii_lowercase() || byte.is_ascii_digit())
    }) {
        return Err(ProtocolError::InvalidLabel);
    }

    Ok(label)
}

pub fn normalize_domain(input: &str) -> Result<String, ProtocolError> {
    if input.is_empty() || input.len() > 253 || !input.is_ascii() || input.ends_with('.') {
        return Err(ProtocolError::InvalidDomain);
    }

    let domain = input.to_ascii_lowercase();
    let valid = domain.split('.').all(|part| {
        !part.is_empty()
            && part.len() <= 63
            && !part.starts_with('-')
            && !part.ends_with('-')
            && part
                .bytes()
                .all(|byte| byte.is_ascii_lowercase() || byte.is_ascii_digit() || byte == b'-')
    });

    if valid {
        Ok(domain)
    } else {
        Err(ProtocolError::InvalidDomain)
    }
}

pub fn generate_tag(
    key: &[u8; KEY_BYTES],
    domain: &str,
    label: &str,
) -> Result<String, ProtocolError> {
    let domain = normalize_domain(domain)?;
    let label = normalize_label(label)?;
    let digest = hmac_digest(key, &domain, &label);
    Ok(encode_tag(&digest[..5]))
}

pub fn generate_address(
    key: &[u8; KEY_BYTES],
    domain: &str,
    label: &str,
) -> Result<String, ProtocolError> {
    let domain = normalize_domain(domain)?;
    let label = normalize_label(label)?;
    let tag = generate_tag(key, &domain, &label)?;
    Ok(format!("{label}-{PROTOCOL_VERSION}-{tag}@{domain}"))
}

pub fn verify_address(
    key: &[u8; KEY_BYTES],
    expected_domain: &str,
    address: &str,
) -> Result<Option<VerifiedAlias>, ProtocolError> {
    let expected_domain = normalize_domain(expected_domain)?;
    if !address.is_ascii() {
        return Ok(None);
    }

    let address = address.to_ascii_lowercase();
    let (local_part, domain) = address
        .rsplit_once('@')
        .ok_or(ProtocolError::InvalidAddress)?;
    if local_part.contains('@') {
        return Ok(None);
    }

    let domain = match normalize_domain(domain) {
        Ok(domain) => domain,
        Err(_) => return Ok(None),
    };
    if domain != expected_domain {
        return Ok(None);
    }

    let (label, tag) = match local_part.rsplit_once("-v1-") {
        Some(parts) => parts,
        None => {
            if looks_like_versioned_alias(local_part) {
                return Err(ProtocolError::UnsupportedVersion);
            }
            return Ok(None);
        }
    };

    let label = match normalize_label(label) {
        Ok(label) => label,
        Err(_) => return Ok(None),
    };
    if tag.len() != TAG_BYTES || !tag.bytes().all(|byte| TAG_ALPHABET.contains(&byte)) {
        return Ok(None);
    }

    let expected_tag = generate_tag(key, &expected_domain, &label)?;
    if bool::from(expected_tag.as_bytes().ct_eq(tag.as_bytes())) {
        Ok(Some(VerifiedAlias {
            address: format!("{label}-{PROTOCOL_VERSION}-{expected_tag}@{expected_domain}"),
            label,
        }))
    } else {
        Ok(None)
    }
}

fn hmac_digest(key: &[u8; KEY_BYTES], domain: &str, label: &str) -> [u8; 32] {
    let mut mac = HmacSha256::new_from_slice(key).expect("HMAC accepts a 32-byte key");
    mac.update(CONTEXT);
    mac.update(&[0]);
    mac.update(domain.as_bytes());
    mac.update(&[0]);
    mac.update(label.as_bytes());
    let digest = mac.finalize().into_bytes();
    let mut output = [0_u8; 32];
    output.copy_from_slice(&digest);
    output
}

fn encode_tag(bytes: &[u8]) -> String {
    debug_assert_eq!(bytes.len(), 5);
    let mut value = 0_u64;
    for &byte in bytes {
        value = (value << 8) | u64::from(byte);
    }

    let mut output = [0_u8; TAG_BYTES];
    for (index, slot) in output.iter_mut().enumerate() {
        let shift = 35 - (index * 5);
        let alphabet_index = ((value >> shift) & 0x1f) as usize;
        *slot = TAG_ALPHABET[alphabet_index];
    }

    String::from_utf8(output.to_vec()).expect("the tag alphabet is ASCII")
}

fn looks_like_versioned_alias(local_part: &str) -> bool {
    let mut parts = local_part.rsplitn(3, '-');
    let _tag = parts.next();
    let version = parts.next();
    let label = parts.next();
    label.is_some() && version.is_some_and(|value| value.starts_with('v'))
}

#[cfg(test)]
mod tests {
    use hmac::{Hmac, Mac};
    use serde::Deserialize;
    use sha2::Sha256;

    use super::*;

    #[derive(Debug, Deserialize)]
    struct TestVectors {
        version: u8,
        vectors: Vec<TestVector>,
    }

    #[derive(Debug, Deserialize)]
    struct TestVector {
        key: String,
        domain: String,
        label: String,
        message_hex: String,
        hmac_sha256: String,
        tag: String,
        address: String,
    }

    #[test]
    fn official_test_vector_matches() {
        let vectors: TestVectors =
            serde_json::from_str(include_str!("../../../test-vectors/v1.json")).unwrap();
        assert_eq!(vectors.version, 1);

        for vector in vectors.vectors {
            let key = decode_key(&vector.key).unwrap();
            assert_eq!(
                generate_tag(&key, &vector.domain, &vector.label).unwrap(),
                vector.tag
            );
            assert_eq!(
                generate_address(&key, &vector.domain, &vector.label).unwrap(),
                vector.address
            );

            let message = format!("mailias/v1\0{}\0{}", vector.domain, vector.label);
            assert_eq!(hex(message.as_bytes()), vector.message_hex);
            let mut mac = Hmac::<Sha256>::new_from_slice(&key).unwrap();
            mac.update(message.as_bytes());
            assert_eq!(hex(&mac.finalize().into_bytes()), vector.hmac_sha256);
        }
    }

    #[test]
    fn labels_are_canonicalized_but_not_rewritten() {
        assert_eq!(normalize_label("GitHub-Work").unwrap(), "github-work");
        assert_eq!(normalize_label("a".repeat(48).as_str()).unwrap().len(), 48);
        assert_eq!(
            normalize_label("a".repeat(49).as_str()),
            Err(ProtocolError::LabelTooLong)
        );

        for invalid in [
            "",
            "-github",
            "github-",
            "github--work",
            "github_work",
            "日本語",
        ] {
            assert!(
                normalize_label(invalid).is_err(),
                "{invalid} should be invalid"
            );
        }
    }

    #[test]
    fn verification_is_case_insensitive_and_domain_bound() {
        let key = [7_u8; KEY_BYTES];
        let address = generate_address(&key, "m.example.test", "github").unwrap();
        let upper = address.to_ascii_uppercase();

        let verified = verify_address(&key, "M.EXAMPLE.TEST", &upper)
            .unwrap()
            .unwrap();
        assert_eq!(verified.label, "github");
        assert_eq!(verified.address, address);
        assert!(verify_address(&key, "example.com", &address)
            .unwrap()
            .is_none());
    }

    #[test]
    fn modified_tag_is_invalid() {
        let key = [9_u8; KEY_BYTES];
        let address = generate_address(&key, "m.example.test", "github").unwrap();
        let modified = address.replacen("-v1-", "-v1-0", 1);
        assert!(verify_address(&key, "m.example.test", &modified)
            .unwrap()
            .is_none());
    }

    fn hex(bytes: &[u8]) -> String {
        bytes.iter().map(|byte| format!("{byte:02x}")).collect()
    }
}
